File: python/teleia/llm.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Iterator
from urllib import error, request

DEFAULT_BASE_URL = "http://127.0.0.1:11434/v1"


@dataclass
class ToolCall:
    id: str
    name: str
    arguments: str


@dataclass
class Message:
    role: str
    content: str | None = None
    tool_calls: list[ToolCall] = field(default_factory=list)
    tool_call_id: str | None = None

    def to_wire(self) -> dict[str, Any]:
        out: dict[str, Any] = {"role": self.role}
        if self.content is not None:
            out["content"] = self.content
        if self.tool_calls:
            out["tool_calls"] = [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {"name": tc.name, "arguments": tc.arguments},
                }
                for tc in self.tool_calls
            ]
        if self.tool_call_id is not None:
            out["tool_call_id"] = self.tool_call_id
        return out


@dataclass
class ContentDelta:
    text: str


@dataclass
class StreamDone:
    tool_calls: list[ToolCall]


StreamEvent = ContentDelta | StreamDone


class LlmClient:
    def __init__(self, base_url: str, model: str) -> None:
        self.base_url = base_url.rstrip("/")
        self.model = model

    def stream(
        self, messages: list[Message], tools: list[dict[str, Any]] | None = None
    ) -> Iterator[StreamEvent]:
        body: dict[str, Any] = {
            "model": self.model,
            "messages": [m.to_wire() for m in messages],
            "stream": True,
        }
        if tools:
            body["tools"] = tools
        data = json.dumps(body).encode("utf-8")
        req = request.Request(
            f"{self.base_url}/chat/completions",
            data=data,
            headers={"content-type": "application/json"},
            method="POST",
        )

        accumulated: dict[int, dict[str, str]] = {}

        try:
            resp = request.urlopen(req, timeout=300)
        except error.HTTPError as e:
            raise RuntimeError(f"ollama {e.code}: {e.read().decode('utf-8', 'replace')}") from e

        with resp:
            for raw_line in resp:
                line = raw_line.decode("utf-8", "replace").rstrip("\n").rstrip("\r")
                if not line.startswith("data:"):
                    continue
                payload = line[5:].strip()
                if not payload or payload == "[DONE]":
                    continue
                try:
                    chunk = json.loads(payload)
                except json.JSONDecodeError:
                    continue
                for choice in chunk.get("choices") or []:
                    delta = choice.get("delta") or {}
                    text = delta.get("content")
                    if text:
                        yield ContentDelta(text)
                    for tcd in delta.get("tool_calls") or []:
                        _accumulate(accumulated, tcd)

        tool_calls = [
            ToolCall(
                id=acc.get("id", ""),
                name=acc.get("name", ""),
                arguments=acc.get("arguments", ""),
            )
            for _, acc in sorted(accumulated.items())
        ]
        yield StreamDone(tool_calls)


def _accumulate(acc: dict[int, dict[str, str]], delta: dict[str, Any]) -> None:
    idx = int(delta.get("index", 0))
    slot = acc.setdefault(idx, {"id": "", "name": "", "arguments": ""})
    if delta.get("id"):
        slot["id"] = delta["id"]
    fn = delta.get("function") or {}
    if fn.get("name"):
        slot["name"] = fn["name"]
    if fn.get("arguments") is not None:
        slot["arguments"] += fn["arguments"]

File: lua/teleia/llm.lua
-- Ollama OpenAI-compat client. Streaming via `curl -N` piped through io.popen.
local json = require("teleia.json")

local M = {}

M.DEFAULT_BASE_URL = "http://127.0.0.1:11434/v1"

local function shell_escape(s)
  return "'" .. tostring(s):gsub("'", "'\\''") .. "'"
end

function M.new(base_url, model)
  return {
    base_url = (base_url:gsub("/+$", "")),
    model = model,
  }
end

-- Iterator returning stream events:
--   { kind = "content", text = "..." }
--   { kind = "done", tool_calls = {...} }
function M.stream(client, messages, tools)
  local body = {
    model = client.model,
    messages = messages,
    stream = true,
  }
  if tools and #tools > 0 then
    body.tools = tools
  end
  local payload = json.encode(body)

  local tmp_in = os.tmpname()
  local fin = assert(io.open(tmp_in, "w"))
  fin:write(payload)
  fin:close()

  local cmd = string.format(
    "curl -fsS -N -X POST -H 'content-type: application/json' --data-binary @%s %s 2>/dev/null",
    shell_escape(tmp_in),
    shell_escape(client.base_url .. "/chat/completions")
  )
  local pipe = assert(io.popen(cmd, "r"))

  local acc = {}
  local indices = {}
  local done = false
  local queue = {}

  local function flush_done()
    table.sort(indices)
    local tool_calls = {}
    for _, idx in ipairs(indices) do
      local slot = acc[idx]
      table.insert(tool_calls, {
        id = slot.id or "",
        ["type"] = "function",
        ["function"] = { name = slot.name or "", arguments = slot.arguments or "" },
      })
    end
    return { kind = "done", tool_calls = tool_calls }
  end

  local function ingest_line(line)
    if not line:match("^data:") then return end
    local raw = line:sub(6):gsub("^%s+", "")
    if raw == "" or raw == "[DONE]" then return end
    local ok, chunk = pcall(json.decode, raw)
    if not ok or not chunk then return end
    for _, choice in ipairs(chunk.choices or {}) do
      local delta = choice.delta or {}
      if delta.content and delta.content ~= "" then
        table.insert(queue, { kind = "content", text = delta.content })
      end
      for _, tcd in ipairs(delta.tool_calls or {}) do
        local idx = tcd.index or 0
        if not acc[idx] then
          acc[idx] = { id = "", name = "", arguments = "" }
          table.insert(indices, idx)
        end
        local slot = acc[idx]
        if tcd.id and tcd.id ~= "" then slot.id = tcd.id end
        local fn = tcd["function"] or {}
        if fn.name and fn.name ~= "" then slot.name = fn.name end
        if fn.arguments then slot.arguments = slot.arguments .. fn.arguments end
      end
    end
  end

  return function()
    while #queue == 0 and not done do
      local line = pipe:read("*l")
      if line == nil then
        done = true
        local close_ok, what, code = pipe:close()
        os.remove(tmp_in)
        if not close_ok then
          table.insert(queue, {
            kind = "content",
            text = string.format("[error: ollama request failed (curl %s %s)]",
              tostring(what or "?"), tostring(code or "?"))
          })
        end
        table.insert(queue, flush_done())
        break
      end
      ingest_line(line)
    end
    if #queue > 0 then
      return table.remove(queue, 1)
    end
    return nil
  end
end

return M
